### ISSUE TITLE
Installing SAP HANA, express edition with Docker - container wont start again (2)

### ISSUE THREAD

**Issue by gsgxnet** [SAPDocuments/Tutorials#2279]
more often than not the container can not be started again. It fails a few seconds after a `docker start DKHXE202`

the startup log:  
```
...

StartService
OK
OK
Starting up HANA ...

18.03.2018 13:16:58
Start
OK

18.03.2018 13:17:10
StartWait
FAIL: process hdbdaemon HDB Daemon not running
```

`docker ps -a` shows for the container:

```
NAMES               IMAGE                                              CREATED ago          STATUS}
DKHXE202            store/saplabs/hanaexpress:2.00.022.00.20171211.1   8 days ago ago       Exited (2) 11 hours ago}
```

in the `trace` dir there is a file 

it contains 2 lines with ` e `:
```
...
[1139]{-1}[-1/-1] 2018-03-18 13:43:37.859636 e Daemon           NetworkListener.cpp(00796) : Failed to resolve empty hostname
...
[1164]{-1}[-1/-1] 2018-03-18 13:43:38.006257 e Daemon           MainImpl.cpp(00340) : Lock file </usr/sap/HXE/HDB90/localhost/lock/hdbdaemon@39000.pid> with pid 1139 already exists. Stop the daemon process
...
```
The first seems to be unimportant, it is there even after a successful run / start. 
But the second seems to be the pointer to the problem:  

and indeed after the exit there still is a `.pid` file in the `locks` dir:  
`hdbdaemon@39000.pid` containing `0000001139`

I wonder what is the right procedure to stop and start a HXE container?  
- A simple `docker stop DKHXE202` had sometimes led to such fails when trying to start it again.  
- Opening a shell within the container and a command `HDB stop` and thereafter a `docker stop` gave a start again fail too.  
- Shutdown and reboot and thereafter a `docker start ...` - same.

**Comment by thecodester** [comment]
@JCurrie27 @aplata-sap Can you take a look here?

**Comment by gsgxnet** [comment]
may be this problem is obsolete now with the new SPS 03? 
on docker store we have now:
```
docker pull store/saplabs/hanaexpress:2.00.030.00.20180403.2
```
my problems happened with the SPS 02. 

If I find some spare time the next days I will give the new docker container image a try.

> Shutdown and reboot and thereafter a docker start ... - same.  

means: shutdown and reboot of the docker host, while HXE container is running.

**Comment by gsgxnet** [comment]
Results so far:

At first SPS 03 (with XSA) looked good. Clean stop can be at least achieved by issuing:
`docker stop -t 320 container`. According to the help this extends the time before the `docker stop` sends a kill signal (=hard stop) to the HXE container from the default 10 secs to 310. Within that time frame a safe HDB shutdown seems to work. 

But I was able to get back to a not any more startable container with a similar  pid exist failure I had posted before:
from the `trc`:  
`Lock file </usr/sap/HXE/HDB90/hxehost/lock/hdbdaemon@39000.pid> with pid 4113 already exists. Stop the daemon process`  
this is the last entry there. 
Seems to be caused by a `deep suspend` (normal suspend but for more than a few hours).
At wake up the container was unresponsive. Tried `xs apps` within. Result minutes of 100% CPU load on all cores, nothing else.  
Had to stop it by a `docker stop -t 300`. Even within that grace time the container did not do a clean shutdown.  
`docker ps -a` shows a 137 exit code for it.  
Restart not possible, see above.

**Comment by gsgxnet** [comment]
sorry closed by error, reopen

**Comment by gsgxnet** [comment]
sorry closed by error, reopen

**Comment by gsgxnet** [comment]
Finally I can state that there are still similar problems with not (re)startable HXE containers where the `.pid` file as leftover from a hard shutdown is the reason for that behavior.

The problem can be reproduced easily:  
either after a longer suspend to RAM or after a hibernate to disk of the host, HXE will run wild (100% CPU on all cores) at wakeup. 

 This unresponsive hdb has to be stopped (killed) hard. 
Restart will fail:  (from the stdout2):
```
An instance of this service is already running.
You should consider checking if the process with the id 413 is running
and of type hdbdaemon. If you are sure that there is no instance alive
remove the file /usr/sap/HXE/HDB90/hxehost/lock/hdbdaemon@39000.pid
and restart.
```

`.pid` file can be removed when the container had been setup with the additional parameter: `--dont-exit-on-error` by a simple `rm /usr/sap/HXE/HDB90/hxehost/lock/hdbdaemon@39000.pid` within the container bash. 
The next start of the same container will (normally) succeed then.  

My suggestions for the tutorials:  
- Describe how to include the optional parameter `--dont-exit-on-error` as way to troubleshoot more easily.  
- in a troubleshooting section mention the removal of the `.pid` file as a (may be) remedy.

Thanks.

**Comment by gsgxnet** [comment]
Question:

should I open a new issue for the running wild HXE here or is there an other way to semi-officially report that problem?

**Comment by gsgxnet** [comment]
the grace period can be extended for an instance when the instance is first created by the `docker run` command. 
Additional parameter should be:  
`--stop-timeout 330`
see: https://docs.docker.com/engine/reference/commandline/run/#stop-container-with-timeout---stop-timeout

With that setup the shutdown sequence of the HDB should have finished within the grace period. No more SIGKILL from the `docker stop` to the HXE container.

if you agree please include this option into the tutorial

**Comment by JCurrie27** [comment]
Hello gsgxnet: the development team is evaluating these stop/start scenarios and we'll respond as soon as we can. In the meantime, please continue monitoring the Community (https://www.sap.com/developer/topics/sap-hana-express.community.html#community) in case other users have found workarounds to this issue.

**Comment by gsgxnet** [comment]
Hi JCurrie27,

I had developed my own workarounds soon after I had posted the scenarios. A bit complicated and a bit difficult to describe and implement - in summary I have now `systemd` stop jobs to safely stop `hdb` in case of system shutdown or system sleep. 

The shutdown `systemd` would not be needed if the docker stop would be set up for the extended grace period, see above. If you think my a bit clumsy  workaround could be useful for others I can post it here.

At the moment I do not see any other option to handle the sleep/wakeup issue other than stopping the HXE before sleep. Could post my workaround too. OTOH I doubt HXE and docker are designed for laptop use.
I do not like to fill issue conversions with one time solutions which are useless for others besides myself. I had hit too many such workarounds which only were a major waste of time in other spaces in the past.
